Feature/identity service validation

DaanTol opened the pull request.

DaanTol committed:
Merge pull request #42 from PTSS-Support/development

Get newest changes

DaanTol committed:
Added identity service validation (Mock)

DaanTol committed:
Some other small change like adding the roles

DaanTol committed:
Improved the Authentication

DaanTol committed:
Merge pull request #49 from PTSS-Support/development

Checking if this works because code doesn't compile

DaanTol committed:
Made some update so only admin can access

SemPlaatsman commented on src/main/kotlin/org/ptss/support/domain/enums/Role.kt:
Waar is de family member en primary caregiver?

DaanTol replied:
Heb identityClient en JwtValidator verbeterd zou je hier even naar kunnen kijken? Ook heb ik de Rollen aangepast en folder CQRS commands genoemd and queries is -> querries geworden

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,52 @@
+    private val jwtValidator: JwtValidator
+) : ContainerRequestFilter {
+
+    override fun filter(requestContext: ContainerRequestContext) {
```
Pls wat documentatie over deze filter. 
e.g.:
* Overwrite een class level filter een function level filter?
* Does a function level filter extend the class level filter?
* Verder voorbeeld: wat als er een class level filter voor admin on the controller staat, en een function level filter voor patient. Kan een admin dan bij die function of alleen de patient? 

Zorg voor duidelijke documentatie en het uitdenken van verschillende cases.

DaanTol replied:
Ben hier nu mee bezig

SemPlaatsman reviewed (changes requested):
Dit geldt voor IdentityServiceClient.kt en JwtValidtor.kt:
Ik heb liever daadwerkelijke JWTs als voorbeeld. De JwtValidator mag namelijk niet gemocked worden en moet compleet functioneel zijn. Dit betekend dat de JwtValidator ook daadwerkelijk Jwt decoding nodig heeft op basis van een Keycloak key (deze key krijg je van de .env, mag nu gwn een random string zijn). Als de JwtValidator dan de Jwt token heeft gedecode kan de role bekeken worden en gekeken worden of de user Unauthorized of niet is. 

So basically de IdentityServiceClients moet daadwerkelijk geldige Jwt token als string terugsturen, en de JwtValidator moet Jwt tokens kunnen decoden op basis van een secret key en de rollen kunnen checken. En onthoud: alleen de IdentityService(Client) mag gemocked worden, hij moet nog steeds de responses terugsturen die je zou verwachten als de IdentityServiceClient wel volledig functioneel was. 

Zie ook mijn andere comments.

SemPlaatsman commented on src/main/kotlin/org/ptss/support/domain/enums/Role.kt:
```diff
@@ -0,0 +1,5 @@
+package org.ptss.support.domain.enums
+
+enum class Role {
+    Admin, Patient, HCP
```
https://stackoverflow.com/questions/3069743/coding-conventions-naming-enums

DaanTol committed:
Added some roles and gave better names, fixed jwtvalidator and identityclient

SemPlaatsman commented on src/main/kotlin/org/ptss/support/domain/interfaces/querries/IQueryHandler.kt:
```diff
@@ -1,4 +1,4 @@
-package org.ptss.support.domain.interfaces.queries
+package org.ptss.support.domain.interfaces.querries
```
Met dit: `En de queries folder heet "querries"`, bedoelde ik dat `querries` de foute benaming is. Queries is de correcte benaming voor meerdere queries ([zie hier](https://simple.wiktionary.org/wiki/queries)).

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,52 @@
+            ?: resourceInfo.resourceClass.getAnnotation(Authentication::class.java)
+
+        if (annotation != null) {
+            val accessToken = requestContext.getHeaderString("Authorization")
```
Zowel de access token, als refresh token, worden opgestuurd als cookie. Gebruik env variabelen voor de namen van deze cookies.

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,52 @@
+            val refreshToken = requestContext.getCookies()["refresh_token"]?.value
+
+            if (accessToken != null && refreshToken != null) {
+                val accessTokenRole = jwtValidator.validateAccessToken(accessToken, annotation.roles)
```
Dit is verkeerde logica. `validateAccessToken` zou niet twee dingen moeten doen. Het zou alleen maar een true of false terug moeten sturen of de token geldig is of niet. Je wilt een extra methode die de rol uit een token kan halen (`extractRole()` of `extractClaim("role")` bijvoorbeeld, tweede misschien iets netter). 

Ook, probeer zoveel mogelijk `var` te vermijden, het maakt de code minder duidelijk, je wilt juist meer "verbose" coderen. 

Verder, de refresh token heeft geen rollen, dat heeft alleen de access token. Alleen de access token bevat ook user informatie. 

Misschien wel het meest belangrjike, je code kan wat opgeschoond worden. Bekijk deze video over [never nesters](https://www.youtube.com/watch?v=CFRhGnuXG-4).

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,52 @@
+) : ContainerRequestFilter {
+
+    override fun filter(requestContext: ContainerRequestContext) {
+        val annotation = resourceInfo.resourceMethod.getAnnotation(Authentication::class.java)
```
Een function-level annotation overschrijft nu een class-level annotation. Het zoekt eerst naar de method-level annotatie, en alleen als die niet bestaat (?:), kijkt het naar de class-level annotatie. Is dit wenselijk?

Ik vindt het goed hoor en een simpel en duidelijk systeem, zolang het maar goed beargumenteerd is.

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/JwtValidator.kt:
```diff
@@ -0,0 +1,45 @@
+            throw UnauthorizedException("Invalid refresh token")
+        }
+    }
+
```
Misschien gwn een validateToken methode? En dan een andere methode die een claim (e.g. role) kan extracten uit een token?

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/JwtValidator.kt:
```diff
@@ -0,0 +1,45 @@
+            "PRIMARY_CAREGIVER" -> Role.PRIMARY_CAREGIVER
+            else -> throw UnauthorizedException("Invalid role in token")
+        }
+    }
```
Aangezien de naam van de Role enum record en de string exact hetzelfde zijn, is het niet makkelijker om de roleClaim proberen te converten naar een Role instance? Maakt de code ook wat meer dynamisch. Hier is wat example code:
```kotlin
private fun mapClaimToRole(roleClaim: String): Role = logger.executeWithExceptionLoggingAsync(
    operation = { Role.valueOf(roleClaim.uppercase()) },
    logMessage = "Invalid role claim: %s",
    exceptionHandling = { UnauthorizedException("Invalid role claim: $roleClaim") },
    roleClaim
)
```

Verder, waarom kan deze functie als input een nullable String hebben (`String?`). Als de roleClaim nullable is, zal het nooit goed zijn, dat weet je ook voordat je hem probeert te converten naar een Role. 

Als laatste, misschien kan je wel een `fromString` functie aan de Role enum toevoegen, maar dat is een persoonlijke stylistic choice. Het voordeel is dat de functie van overal gebruikt kan worden, maar je wilt dan misschien weer geen logging in the Role enum.

SemPlaatsman reviewed (changes requested):
Heb wat comments gemaakt. 

[Dit is misschien wel het meest belangrijke](https://www.youtube.com/watch?v=CFRhGnuXG-4)

DaanTol committed:
Gave queries correct name, added mapper and constant,  cleaned up code, added md

Also looked at all comments of reviewer and implemented the changes requested, use of "verbode" and no "var" better logic, a env variable and no nesting

SemPlaatsman commented on AUTHFILTER.md:
Hernoem deze file naar `README.md` en stop hem in de `security` folder

SemPlaatsman commented on src/main/kotlin/org/ptss/support/api/controllers/ProductController.kt:
```diff
@@ -12,18 +12,23 @@ import org.ptss.support.api.dtos.requests.CreateProductRequest
     override suspend fun getAllProducts(): List<ProductResponse> =
         productFacade.getAllProducts()
 
+    @Authentication(roles = [Role.PATIENT], message = "To get a product, you should be authenticated as an admin")
```
Message komt niet overeen met role

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/JwtValidator.kt:
```diff
@@ -0,0 +1,33 @@
+class JwtValidator @Inject constructor(
+    private val jwtParser: JWTParser
+) {
+    fun isAccessTokenValid(accessToken: String, allowedRoles: Array<Role>): Boolean =
```
Dit wil je niet hier doen, maar in je AuthenticationFilter. Het is niet de rol van de JwtValidator om te zeggen of de toegestane rollen overeenkomen met de rol in de token. Verwijder deze methode en stop de logica in de AuthenticationFilter.

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,75 @@
+            jwtValidator.isAccessTokenValid(accessToken, annotation.roles) -> return
+            !refreshToken.isNullOrBlank() -> handleRefreshToken(requestContext, refreshToken, annotation)
+            else -> denyRequest(requestContext, annotation.message)
+        }
```
Dit is eigenlijk verkeerde logica, want als de access token valid is maar de refresh token niet, dan kan je nog steeds authenticaten. Verder, uit de never nesting video die ik had gestuurd staat vlgns mij ook dat je niet wilt dat je happy flow middenin de code staat, maar bijna altijd aan het einde van je code. 

Dit zijn de checks die ik voorstel:
[Hier is een artifact met verbeterde code als voorbeeld](https://claude.site/artifacts/688db41a-99a3-4b37-971d-f371e5310063)

SemPlaatsman commented on src/main/kotlin/org/ptss/support/core/mappers/RoleMapper.kt:
```diff
@@ -0,0 +1,11 @@
+import io.quarkus.security.UnauthorizedException
+import org.ptss.support.domain.enums.Role
+
+object RoleMapper {
```
Liever niet een hele nieuwe mapper voor één use case. Zorg ervoor dat er een fromString function in the Role enum class zit en gebruik die.

SemPlaatsman reviewed (changes requested):
Veel fouten in de logica. 

Meest belangrijke informatie staat [hierin](https://claude.site/artifacts/688db41a-99a3-4b37-971d-f371e5310063). Verder graag geen mapper voor de enkele purpose van het vertalen van claim naar Role gebruiken, een mapper hoort voornamelijk gebruikt te woorden voor het vertalen van en naar DTOs en domain models. 

Verder hele goeie documentatie voor de auth filter, nog even op de goede plaats zetten en dan is dat stukje toppie. 

Adjust accordingly

DaanTol committed:
Moved doc to security, removed isAccessTokenValid in JWTvalidtor placing it ine authfilter,removed  mapper, added fromstring to Role

DaanTol committed:
Logic no longer gives a 500 error

SemPlaatsman commented on src/main/kotlin/org/ptss/support/domain/enums/Role.kt:
```diff
@@ -0,0 +1,15 @@
+            return role?.takeUnless { it.isBlank() }
+                ?.let { valueOf(it.uppercase()) }
+                ?: throw UnauthorizedException("Role claim is missing or invalid")
+        }
```
Dit is niet goed, je moet rekening houden met de rol die de Role enum heeft, en dat is, een enum zijn. De Role enum zou geen belang of kennis moeten hebben van claims logica of dat een user unauthorized is als het geen geldige rol input. 

Daarbovenop, je wilt niet `fromString(role: String?)` (nullable string) hebben maar `fromString(role: String)` (non-nullable string). Het zou namelijk niet logisch zijn als je Role.fromString aanroept zonder een daadwerkelijke string te geven. Je zou dus eigenlijk meer zoiets willen:
```kotlin
fun fromString(value: String): Role {
    return try {
        valueOf(value.uppercase())
    } catch (e: IllegalArgumentException) {
        throw IllegalArgumentException("Invalid role: $value. Valid values are: ${values().joinToString()}")
    }
}
```
of korter (stylistic choice):
```kotlin
fun fromString(value: String): Role = 
    runCatching { valueOf(value.uppercase()) }
        .getOrElse { throw IllegalArgumentException("Invalid role: $value. Valid values are: ${values().joinToString()}") }
```
Adjust accordingly.

DaanTol replied:
Dus zo?

"enum class Role {
    ADMIN, PATIENT, HCP, FAMILY_MEMBER, PRIMARY_CAREGIVER;

    companion object {
        fun fromString(value: String): Role =
            runCatching { valueOf(value.uppercase()) }
                .getOrElse { throw IllegalArgumentException("Invalid role: $value. Valid values are: ${values().joinToString()}") }
    }
}"

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,90 @@
+        val refreshToken = requestContext.cookies[refreshTokenCookieName]?.value
+
+        // Deny if refresh token is missing or invalid
+        if (refreshToken.isNullOrBlank() || !jwtValidator.isTokenValidAndNotBlank(refreshToken)) {
```
`refreshToken.isNullOrBlank() || ` kan nu weggehaald worden omdat dat al gecheckt worden door `isTokenValidAndNotBlank`.

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,90 @@
+            refreshToken.isNullOrBlank() -> {
+                denyRequest(requestContext, "Missing refresh token")
+                return
+            }
```
Dit:
```kotlin
refreshToken.isNullOrBlank() -> {
    denyRequest(requestContext, "Missing refresh token")
    return
}
```
kan weg omdat je door je eerdere check (`!jwtValidator.isTokenValidAndNotBlank(refreshToken)`) al weet dat de `refreshToken` valid is.

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,90 @@
+        newToken
+    }.getOrElse {
+        denyRequest(requestContext, "Unauthorized")
+        throw UnauthorizedException("Failed to refresh token")
```
Is het nog nodig om een UnauthorizedException te gooien als je al `denyRequest(...)` hebt aangeroepen?
Ik vraag het omdat ik dit niet weet, ik was me ervan bewust dat `denyRequest` gelijk de code stopt en een Unauthorized message terugstuurt. In de vorige code doe je namelijk wel `denyRequest` en daarna ga je ervan uit dat de code daarna niet meer wordt uitgevoerd.

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,90 @@
+                .entity(message)
+                .build()
+        )
+    }
```
Als ik het goed begrijp stuurt dit een 401 Unauthorized terug met een message:
```
401 Unauthorized
{
   "message": "the provided message"
}
```
Dit volgt helaas niet de `APIException` of `createResponse` structure. Je wilt deze structuur aanhouden. Daarbovenop, als jij een Exception gooit en niet vangt vanuit de `AuthenticationFilter`, wordt die dan opgevangen door de `GlobalExceptionHandler`? Pls test dit want als dit zo is kan je gwn een Exception gooien ipv aborten. 

In kort, zorg ervoor dat dit clean en gestandaardiseerd is met zo min mogelijk duplicate code.

DaanTol committed:
Deny request was not needed at all and ugly, replaced String? with String and removed other useless checks

Tayam-El-Jebari commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,79 @@
+
+        // Validate refresh token
+        if (!jwtValidator.isTokenValidAndNotBlank(refreshToken)) {
+            throw UnauthorizedException(annotation.message)
```
why not more info in the exception? Is this UnauthorizedException being catched by the global exception handler? If not, please still do?

SemPlaatsman replied:
Ik zie nu ook dat er in de `GlobalExceptionHandler` geen record is voor een `UnauthorizedException`, dit moet dan ook toegevoegd worden als dit door de `GlobalExceptionHandler` opgevangen wordt. 

Over het stukje van meer info in de Exception: het hangt ervan af: wordt het naar de user opgestuurd of niet? Want je wilt de user niet laten weten dat de refreshToken niet geldig is, dat is niet hun concern.

Tayam-El-Jebari replied:
Ligt er idd aan hoe de aanpassing op globalexceptionhandler eruit gaat zien

Tayam-El-Jebari commented on src/main/kotlin/org/ptss/support/security/JwtValidator.kt:
```diff
@@ -0,0 +1,32 @@
+package org.ptss.support.security
+
+import io.quarkus.security.UnauthorizedException
```
don't think this import is being used

Tayam-El-Jebari commented on src/main/kotlin/org/ptss/support/security/IdentityServiceClient.kt:
```diff
@@ -0,0 +1,49 @@
+) {
+    fun refreshAccessToken(refreshToken: String): String {
+        val newToken = getNewAccessToken(refreshToken)
+        if (!jwtValidator.isTokenValidAndNotBlank(newToken)) {
+            throw UnauthorizedException("Failed to obtain valid access token")
+        }
```
would throw here our custom apiexception, as it isn't really a unauthorized exception
```
        throw APIException(
            errorCode = ErrorCode.TOKEN_GENERATION_FAILED,
            message = "Failed to generate valid access token"
        )
```

Tayam-El-Jebari commented on src/main/kotlin/org/ptss/support/security/JwtValidator.kt:
```diff
@@ -0,0 +1,32 @@
+class JwtValidator @Inject constructor(
+    private val jwtParser: JWTParser
+) {
+    fun isTokenValidAndNotBlank(token: String?): Boolean =
+        !token.isNullOrBlank() && isTokenValid(token)
```
not sure how this eventually would work. Do the refresh token and accesstoken use the same validatetoken method?

SemPlaatsman replied:
Yes, both are JWT and validated against a Keycloak Key

Tayam-El-Jebari commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,79 @@
+    @Context private val resourceInfo: ResourceInfo,
+    private val identityServiceClient: IdentityServiceClient,
+    private val jwtValidator: JwtValidator,
+    @ConfigProperty(name = "app.security.access-token-cookie-name", defaultValue = "access_token")
+    private val accessTokenCookieName: String,
+    @ConfigProperty(name = "app.security.refresh-token-cookie-name", defaultValue = "refresh_token")
+    private val refreshTokenCookieName: String
```
These should come from your requests right?  Why do they come from application.properties instead?
would be fine if you did this, no?
`val accessToken = requestContext.cookies[accessTokenCookieName]?.value`

SemPlaatsman replied:
What do you mean come from requests, how can you know from a request what the name of the refresh token and access token is?

This was mainly done to standardize the naming for refresh and access cookies, because they come from application.properties, they can be updated organization-wide.

Tayam-El-Jebari replied:
ah i see, nvm. Misunderstood the function of this line. Thought the token was retrieved from the application context o_0

I blame sleep deprivation

DaanTol committed:
Changed wrong nam of folder, add UnauthorizedException in GlobalExceptionHandler

Furthermore Gave both refreshAccessToken and getNewAccessToken a APIException and removed not needed imports and added new error code

DaanTol committed:
Made a constant instead

SemPlaatsman commented on src/main/resources/application.properties:
```diff
@@ -35,6 +35,10 @@ quarkus.openshift.resources.requests.cpu=250m
 
+# Cookie Configuration
+app.security.access-token-cookie-name=access_token
+app.security.refresh-token-cookie-name=refresh_token
```
Oei nee, moet uit de .env gehaald worden. En pls een betere naam verzinnen

DaanTol replied:
Waar wil je het dan hebben?

SemPlaatsman commented on src/main/kotlin/org/ptss/support/security/AuthenticationFilter.kt:
```diff
@@ -0,0 +1,82 @@
+            .path("/")
+            .httpOnly(true)
+            .secure(true)
+            .build()
```
Cookie building should be done through .env values, de volgende values moeten door env values set kunnen worden:
```
HttpOnly; 
Secure; 
SameSite=Strict; 
Domain=example.com; 
Path=/; 
Max-Age=3600
```
Ik heb een voordeel van de waardes gegeven

DaanTol replied:
Ging jij dit nou doen?

SemPlaatsman replied:
Ja

DaanTol committed:
Added Sem changes

DaanTol committed:
Added securityproperties file, cleaned up auth filter

DaanTol committed:
Merge branch 'development' into feature/identity-service-validation

DaanTol committed:
Removed a unused import

DaanTol committed:
Merge branch 'feature/identity-service-validation' of https://github.com/Tayam-El-Jebari/PTSS-Support-Quarkus-Base-Repo into feature/identity-service-validation

DaanTol committed:
Testing

DaanTol committed:
Removed .

DaanTol committed:
Was not refering to correct packages

DaanTol committed:
Update to ProductServiceTest
